server/report.go

```diff
@@ -0,0 +1,332 @@
+		linesWritten++
+	}
+
+	if linesWritten == 0 {
+		b.WriteString("No hosts to show.")
+	}
+
+	return b.String()
+}
+
+func formatServiceCount(count nagios.ServiceCount) string {
+	if count.Result.TypeText != resultTypeTextSuccess {
+		return gettingReportUnsuccessfulMessage("service summary", count.Result.TypeText)
+	}
+
+	var b strings.Builder
+
+	b.WriteString("##### SERVICE SUMMARY\n\n")
+	b.WriteString(fmt.Sprintf("%s OK: **%d**", whiteCheckMarkEmoji, count.Data.Count.Ok))
+	b.WriteString(fmt.Sprintf("  %s Warning: **%d**", warningEmoji, count.Data.Count.Warning))
+	b.WriteString(fmt.Sprintf("  %s Critical: **%d**", bangBangEmoji, count.Data.Count.Critical))
+	b.WriteString(fmt.Sprintf("  %s Unknown: **%d**", questionEmoji, count.Data.Count.Unknown))
+	b.WriteString(fmt.Sprintf("  %s Pending: **%d**", hourglassFlowingSandEmoji, count.Data.Count.Pending))
+
+	return b.String()
+}
+
+type extractedService struct {
+	name, state string
+}
+
+// extractServices returns a slice of extractedService, extracted from
+// rawMessage. It returns a slice with single element initialized to a
+// unknownState state if it fails to process rawMessage.
+func extractServices(rawMessage json.RawMessage) []extractedService {
+	var rawStates map[string]json.RawMessage
+
+	if err := json.Unmarshal(rawMessage, &rawStates); err != nil {
+		return []extractedService{{state: unknownState}}
+	}
+
+	var services []extractedService
+
+	for k, v := range rawStates {
+		var state string
+
+		if err := json.Unmarshal(v, &state); err != nil {
+			state = unknownState
+		}
+
+		services = append(services, extractedService{name: k, state: state})
+	}
+
+	return services
+}
+
+func formatServiceList(list nagios.ServiceList) string {
+	if list.Result.TypeText != resultTypeTextSuccess {
+		return gettingReportUnsuccessfulMessage("service list", list.Result.TypeText)
+	}
+
+	var reportLength int
+
+	hostToServices := make(map[string][]extractedService)
+
+	for k, v := range list.Data.ServiceList {
+		services := extractServices(v)
+
+		reportLength += len(services) + 1 // add 1 for a line with hostname.
+
+		hostToServices[k] = services
+	}
+
+	var b strings.Builder
+
+	b.WriteString("##### SERVICE LIST\n\n")
+
+	var abnormalOnly bool
+
+	if reportLength >= maximumReportLength {
+		abnormalOnly = true
+		b.WriteString("**Too many services. Showing only abnormal state services.**\n\n")
+	}
+
+	var linesWritten int
+
+	const theEnd = "\n\n**Skipped the rest of the services.**"
+
+	for host, services := range hostToServices {
+		var hostWritten bool
+		for _, s := range services {
+			if s.state == okState && abnormalOnly {
+				continue
+			}
+
+			if !hostWritten {
+				hostWritten = true
+
+				if linesWritten > 0 {
+					b.WriteRune('\n')
+				}
+				b.WriteString(fmt.Sprintf("`%s`:", host))
+				linesWritten++
+
+				if linesWritten == maximumReportLength {
+					b.WriteString(theEnd)
+					return b.String()
+				}
```

Comment:
Shouldn't we perform this check first, to avoid the possibility of displaying an ambiguous line?

Reply:
Which ambiguous line do you mean?

Reply:
We have talked about it on a chat, so I think I will just mark it resolved.